File: test_suite/testHarness.php
<?php

// Define a constant for the User Agent string
const UASTRING = 'GS1_Resolver_Test/0.1';

// Create a new instance of the Guzzle HTTP client (equivalent to LWP::UserAgent)
$client = new \GuzzleHttp\Client(['headers' => ['User-Agent' => UASTRING]]);

// Define an array to store the request headers
$reqHeaders = [];

// Check for environment variables and add them to the request headers if set
if (!empty($_SERVER['HTTP_USER_AGENT'])) {
    $reqHeaders['User-Agent'] = $_SERVER['HTTP_USER_AGENT'];
}
if (!empty($_SERVER['HTTP_ACCEPT'])) {
    $reqHeaders['Accept'] = $_SERVER['HTTP_ACCEPT'];
}
if (!empty($_SERVER['HTTP_ACCEPT_LANGUAGE'])) {
    $reqHeaders['Accept-Language'] = $_SERVER['HTTP_ACCEPT_LANGUAGE'];
}

// Define a function to parse the URL query string
function parseQueryString($queryString) {
    $params = [];
    foreach (explode('&', $queryString) as $pair) {
        list($key, $value) = explode('=', $pair);
        $params[$key] = urldecode($value);
    }
    return $params;
}

// Define a function to get the HTTP version
function getHTTPVersion($domain) {
    $uri = "http://$domain";
    $response = $client->head($uri, ['headers' => $reqHeaders]);
    $statusCode = $response->getStatusCode();
    if ($statusCode === 200) {
        $versionRegex = '/HTTP\/(\d\.\d).*/';
        preg_match($versionRegex, (string)$response->getBody(), $matches);
        return $matches[1];
    } else {
        throw new Exception("Error getting HTTP version: " . $response->getReasonPhrase());
    }
}

// Define a function to get all headers
function getAllHeaders($uri) {
    $response = $client->head($uri, ['headers' => $reqHeaders]);
    $statusCode = $response->getStatusCode();
    if ($statusCode === 200) {
        $headers = $response->getHeaders();
        $text = '{"httpCode":"' . $statusCode . '",';
        $text .= '"httpMsg":"' . $response->getReasonPhrase() . '",';
        $text .= '"requestURI":"' . $_SERVER['QUERY_STRING'] . '",';
        $text .= '"Requesting_User_Agent":"' . $_SERVER['HTTP_USER_AGENT'] . '",';
        $text .= '"Requesting_Accept_Header":"' . $_SERVER['HTTP_ACCEPT'] . '",';
        $text .= '"Requesting_Accept_Language":"' . $_SERVER['HTTP_ACCEPT_LANGUAGE'] . '",';
        foreach ($headers as $header => $value) {
            if (is_array($value)) {
                $text .= '"' . $header . '":[';
                foreach ($value as $val) {
                    $val = str_replace('"', '\\"', $val);
                    $text .= '"' . $val . '",';
                }
                $text = rtrim($text, ',') . '],';
            } else {
                $value = str_replace('"', '\\"', $value);
                $text .= '"' . $header . '":"' . $value . '",';
            }
        }
        return rtrim($text, ',') . '}';
    } else {
        throw new Exception("Error getting headers: " . $response->getReasonPhrase());
    }
}

// Define a function to show all headers (for debugging)
function showAllHeaders($uri) {
    $response = $client->head($uri, ['headers' => $reqHeaders]);
    $statusCode = $response->getStatusCode();
    if ($statusCode === 200) {
        $headers = $response->getHeaders();
        $text = "Headers for $uri are:\n\n";
        foreach ($headers as $header => $value) {
            $text .= "$header=$value\n\n";
        }
        return $text;
    } else {
        throw new Exception("Error getting headers: " . $response->getReasonPhrase());
    }
}

// Output the HTTP version
echo getHTTPVersion('id.gs1.org') . "\n";

// Output all headers
echo getAllHeaders('https://id.gs1.org') . "\n";

// Output debug information
echo showAllHeaders('https://id.gs1.org') . "\n";

File: test_suite/testHarness.pl
#!/usr/bin/perl
use cPanelUserConfig;
use CGI::Carp qw(fatalsToBrowser);
use strict;
use utf8;
use LWP::UserAgent;

############################################################################
# Functions for GS1 Digital Link Test Suite
############################################################################

######### We're going to need a User Agent
use constant UAstring => 'GS1_Resolver_Test/0.1';
my $browser= LWP::UserAgent-> new();
$browser->requests_redirectable([]); #(suppress all redirects, we want the resolver's response, not the target's)

# We want to pass on the request headers received from the calling application
# Make no assumptions about what the values should be. If no value is set by the client, don't set it here (so default is null).
my @reqHeaders = ();
if ($ENV{'HTTP_USER_AGENT'}) {
  push(@reqHeaders, 'User-Agent' => $ENV{'HTTP_USER_AGENT'});
}
if ($ENV{'HTTP_ACCEPT'}) {
  push(@reqHeaders, 'Accept' => $ENV{'HTTP_ACCEPT'});
}
if ($ENV{'HTTP_ACCEPT_LANGUAGE'}) {
  push(@reqHeaders, 'Accept-language' => $ENV{'HTTP_ACCEPT_LANGUAGE'});
}

my %h = &parseQuery($ENV{'QUERY_STRING'});

my $text;

print "Content-type: application/json\n\n";

if ($h{'test'}) { # So we have a test to carry out
  if (($h{'test'} eq 'getHTTPversion') && ($h{'testVal'})) {
    $text = '{"test":"'.$h{'test'}.'","testVal":"'.$h{'testVal'}.'","result":"';
    $text .= getHTTPversion($h{'testVal'});
    $text .= '"}';
  } elsif (($h{'test'} eq 'getAllHeaders') && ($h{'testVal'})) {
    $text = '{"test":"'.$h{'test'}.'","testVal":"'.$h{'testVal'}.'","result":';
    $text .= getAllHeaders($h{'testVal'});
    $text =~ s/(.*),$/$1/;  # Remove final comma
    $text .= '}}';
  }
} else {
    $text .= "No command received\n";
}


print $text;


## Helper functions ##

sub parseQuery {
  my $query = $_[0];
  my @pairs = split('&', $query);
  my $name;
  my $value;
  my @h;
  foreach (@pairs) {
    ($name, $value) = split(/=/, $_);
    push (@h, $name, url_decode($value));
  }
  return @h;
}


sub url_decode {
  my $r = shift;
  $r =~ tr/\+/ /;
  $r =~ s/%([a-f0-9][a-f0-9])/chr( hex( $1 ) )/gei;
  return $r;
}

sub getHTTPversion {
  $_ = shift;
  my $domain = "http://".$_;
  my $curlResponse = `curl -I $domain`;
  $curlResponse =~ /HTTP\/(\d\.\d).*/;
  return $1;
}

sub getAllHeaders {
  my $uri = shift;
#  my $accept = shift;
#  my $acceptLang = shift;

  my $response = $browser->head($uri, @reqHeaders);		# Do head request
#  die "Hmm, error \"", $response->status_line(), "\" when getting $uri" unless $response->is_success(); # We do't actually want to fail on non 2xx codes
  my @headersObject = $response->headers();
  my $text = "{";
  $text .= '"httpCode":"'.$response->code().'",';   # We want the HTTP codes in our result object, even though they're not actual HTTP headers
  $text .= '"httpMsg":"'.$response->message().'",';
  $text .= '"requestURI":"'.$ENV{'QUERY_STRING'}.'",';
  $text .= '"Requesting_User_Agent":"'.$ENV{'HTTP_USER_AGENT'}.'",';
  $text .= '"Requesting_Accept_Header":"'.$ENV{'HTTP_ACCEPT'}.'",';
  $text .= '"Requesting_Accept_Language":"'.$ENV{'HTTP_ACCEPT_LANGUAGE'}.'",';
  foreach my $h (@headersObject) {
    for my $header (keys %$h) {
      if (ref($h->{$header}) eq 'ARRAY') {  # This happens if the server sends multiple headers of the same type, which seems to be allowed :-(
        $text .= '"'.$header.'":';
        $text .= '[';
        my $i = 0;
        while ($h->{$header}[$i] ne '') {
          $h->{$header}[$i] =~s/"/\\"/g;    # espace any " characters
          $text .= '"'.$h->{$header}[$i].'",';
          $i++;
        }
        $text =~ s/(.*),$/$1/;    # removes final comma
        $text .= '],';
      } else {
        $h->{$header} =~s/"/\\"/g;
        $text .= '"'.$header.'":"'.$h->{$header}.'",';
      }
    }
  }
  return $text;
}


###########################################


# A little function useful for debugging

sub showAllHeaders {
  my $uri = shift;
  my $response = $browser->head($uri, @reqHeaders);		# Do head request
  die "Hmm, error \"", $response->status_line(), "\" when getting $uri" unless $response->is_success();
  my @headersObject = $response->headers();
  my $text = "Headers for $uri are:\n\n";
  foreach my $h (@headersObject) {
    for my $header (keys %$h) {
      $text .= "$header=$h->{$header}\n\n";
    }
  }
  return $text;
}
